File: Trabalho 3 API/Trabalho_3/clientes/js/cliente.js
const axios = require('axios');
const readline = require('readline');

const BASE_URL = "http://172.25.210.208:8080/api";

// Configuração da interface de leitura
const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

// Função para exibir o menu
function mostrarMenu() {
  console.log(`
  === MENU API DE APICULTURA ===
  1. Criar nova colmeia
  2. Listar colmeias
  3. Adicionar abelhas a uma colmeia
  4. Remover colmeia
  5. Sair
  `);
}

// Função principal
async function main() {
  let continuar = true;
  
  while (continuar) {
    mostrarMenu();
    
    const opcao = await perguntar("Digite uma opção: ");
    
    switch(opcao) {
      case '1':
        await criarColmeiaMenu();
        break;
      case '2':
        await listarColmeiasMenu();
        break;
      case '3':
        await adicionarAbelhasMenu();
        break;
      case '4':
        await removerColmeiaMenu();
        break;
      case '5':
        continuar = false;
        break;
      default:
        console.log("Opção inválida!");
    }
  }
  
  rl.close();
}

// Funções auxiliares para entrada
function perguntar(pergunta) {
  return new Promise(resolve => {
    rl.question(pergunta, resolve);
  });
}

// Menus específicos
async function criarColmeiaMenu() {
  console.log("\n--- Criar Nova Colmeia ---");
  const nome = await perguntar("Nome do apicultor: ");
  const abelhas = await perguntar("Capacidade de abelhas: ");
  const mel = await perguntar("Capacidade de mel: ");
  
  const colmeia = await criarColmeia(nome, parseInt(abelhas), parseInt(mel));
  console.log("Colmeia criada:", colmeia);
}

async function listarColmeiasMenu() {
  console.log("\n--- Listar Colmeias ---");
  const nome = await perguntar("Nome do apicultor: ");
  const colmeias = await listarColmeias(nome);
  console.log("Colmeias:", colmeias);
}

async function adicionarAbelhasMenu() {
  console.log("\n--- Adicionar Abelhas ---");
  const id = await perguntar("ID da colmeia: ");
  const quantidade = await perguntar("Quantidade de abelhas: ");
  const rainha = await perguntar("Tem rainha? (s/n): ");
  
  const resultado = await adicionarAbelha(
    parseInt(id),
    parseInt(quantidade),
    rainha.toLowerCase() === 's'
  );
  console.log("Resultado:", resultado);
}

async function removerColmeiaMenu() {
  console.log("\n--- Remover Colmeia ---");
  const id = await perguntar("ID da colmeia a remover: ");
  const resultado = await removerColmeia(parseInt(id));
  console.log("Resultado:", resultado);
}

// Suas funções originais (mantidas sem alteração)
async function criarColmeia(nomeApicultor, capacidadeAbelhas, capacidadeMel) {
  try {
    const response = await axios.post(`${BASE_URL}/colmeia_new`, null, {
      params: { nomeApicultor, capacidadeAbelhas, capacidadeMel }
    });
    return response.data;
  } catch (error) {
    console.error("Erro ao criar colmeia:", error.message);
    if (error.response) {
      console.error("Detalhes:", error.response.data);
    }
  }
}

async function listarColmeias(nomeApicultor) {
  try {
    const response = await axios.get(`${BASE_URL}/colmeia_list`, {
      params: { nomeApicultor }
    });
    return response.data;
  } catch (error) {
    console.error("Erro ao listar colmeias:", error.message);
    if (error.response) {
      console.error("Detalhes:", error.response.data);
    }
  }
}

async function adicionarAbelha(idColmeia, quantidade, rainhaPresente) {
  try {
    const response = await axios.post(`${BASE_URL}/colmeia_add_abelhas`, null, {
      params: { idColmeia, quantidade, rainhaPresente }
    });
    return response.data;
  } catch (error) {
    console.error("Erro ao adicionar abelha:", error.message);
    if (error.response) {
      console.error("Detalhes:", error.response.data);
    }
  }
}

async function removerColmeia(idColmeia) {
  try {
    const response = await axios.delete(`${BASE_URL}/colmeia_del/${idColmeia}`);
    return response.data || "Colmeia removida com sucesso";
  } catch (error) {
    console.error("Erro ao remover colmeia:", error.message);
    if (error.response) {
      console.error("Detalhes:", error.response.data);
    }
  }
}

// Inicia o programa
main().catch(err => console.error("Erro no sistema:", err));

File: Trabalho 3 API/Trabalho_3/clientes/c/cliente.c
#include <stdio.h>
#include <stdbool.h>
#include <curl/curl.h>
#include <string.h>

#define API_URL "http://172.25.210.208:8080/api"

size_t write_callback(void *contents, size_t size, size_t nmemb, void *userp) {
    size_t total_size = size * nmemb;
    printf("%.*s", (int)total_size, (char *)contents);
    return total_size;
}

void criar_colmeia(const char *nome_apicultor, int capacidade_abelhas, int capacidade_mel) {
    CURL *curl = curl_easy_init();
    if (curl) {
        char url[256];
        snprintf(url, sizeof(url), "%s/colmeia_new?nomeApicultor=%s&capacidadeAbelhas=%d&capacidadeMel=%d",
                 API_URL, nome_apicultor, capacidade_abelhas, capacidade_mel);

        curl_easy_setopt(curl, CURLOPT_URL, url);
        curl_easy_setopt(curl, CURLOPT_POST, 1L);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
        curl_easy_setopt(curl, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, "");
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);

        struct curl_slist *headers = NULL;
        headers = curl_slist_append(headers, "Expect:");
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

        printf("{Enviando POST para criar colmeia...}\n");
        CURLcode res = curl_easy_perform(curl);
        if (res != CURLE_OK) {
            fprintf(stderr, "{Erro ao criar colmeia: %s}\n", curl_easy_strerror(res));
        }
        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);
    }
}

void listar_colmeias(const char *nome_apicultor) {
    CURL *curl = curl_easy_init();
    if (curl) {
        char url[256];
        snprintf(url, sizeof(url), "%s/colmeia_list?nomeApicultor=%s", API_URL, nome_apicultor);

        curl_easy_setopt(curl, CURLOPT_URL, url);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);

        printf("{Enviando GET para listar colmeias...}\n");
        CURLcode res = curl_easy_perform(curl);
        if (res != CURLE_OK) {
            fprintf(stderr, "{Erro ao listar colmeias: %s}\n", curl_easy_strerror(res));
        }
        curl_easy_cleanup(curl);
    }
}

void adicionar_abelha(int id_colmeia, int quantidade, bool rainha_presente) {
    CURL *curl = curl_easy_init();
    if (curl) {
        char url[256];
        const char *rainha_str = rainha_presente ? "true" : "false";
        snprintf(url, sizeof(url), "%s/colmeia_add_abelhas?idColmeia=%d&quantidade=%d&rainhaPresente=%s",
                 API_URL, id_colmeia, quantidade, rainha_str);

        curl_easy_setopt(curl, CURLOPT_URL, url);
        curl_easy_setopt(curl, CURLOPT_POST, 1L);
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
        curl_easy_setopt(curl, CURLOPT_HTTP_VERSION, CURL_HTTP_VERSION_1_1);
        curl_easy_setopt(curl, CURLOPT_POSTFIELDS, "");
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);

        struct curl_slist *headers = NULL;
        headers = curl_slist_append(headers, "Expect:");
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

        printf("{Enviando POST para adicionar abelhas...}\n");
        CURLcode res = curl_easy_perform(curl);
        if (res != CURLE_OK) {
            fprintf(stderr, "{Erro ao adicionar abelhas: %s}\n", curl_easy_strerror(res));
        }
        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);
    }
}

void remover_colmeia(int id_colmeia) {
    CURL *curl = curl_easy_init();
    if (curl) {
        char url[256];
        // Corrigido o endpoint para usar path parameter (alinhado com o Spring Boot)
        snprintf(url, sizeof(url), "%s/colmeia_del/%d", API_URL, id_colmeia);

        // Configurações do CURL
        curl_easy_setopt(curl, CURLOPT_URL, url);
        curl_easy_setopt(curl, CURLOPT_CUSTOMREQUEST, "DELETE");
        curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, write_callback);
        curl_easy_setopt(curl, CURLOPT_TIMEOUT, 5L);
        
        // Headers (opcional, mas recomendado)
        struct curl_slist *headers = NULL;
        headers = curl_slist_append(headers, "Content-Type: application/json");
        curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

        printf("Enviando DELETE para remover colmeia ID %d...\n", id_colmeia);
        CURLcode res = curl_easy_perform(curl);
        
        // Tratamento de erros aprimorado
        if (res != CURLE_OK) {
            fprintf(stderr, "Erro na requisição: %s\n", curl_easy_strerror(res));
        } else {
            long http_code = 0;
            curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &http_code);
            
            if (http_code == 200) {
                printf("Colmeia %d removida com sucesso!\n", id_colmeia);
            } else {
                fprintf(stderr, "Falha ao remover. HTTP Status: %ld\n", http_code);
            }
        }
        
        // Limpeza
        curl_slist_free_all(headers);
        curl_easy_cleanup(curl);
    } else {
        fprintf(stderr, "Falha ao inicializar libcurl.\n");
    }
}

void menu() {
    printf("\n{ MENU }\n");
    printf("{1} Criar colmeia\n");
    printf("{2} Listar colmeias\n");
    printf("{3} Adicionar abelhas\n");
    printf("{4} Remover colmeia\n");
    printf("{0} Sair\n");
}

int main() {
    curl_global_init(CURL_GLOBAL_ALL);

    int escolha;
    while (1) {
        menu();
        printf("{Escolha uma opção}: ");
        if (scanf("%d", &escolha) != 1) break;
        getchar(); // Limpa o buffer do teclado

        if (escolha == 1) {
            char nome[100];
            int cap_abelhas, cap_mel;
            printf("{Nome do apicultor}: ");
            fgets(nome, sizeof(nome), stdin);
            nome[strcspn(nome, "\n")] = 0;
            printf("{Capacidade de abelhas}: ");
            scanf("%d", &cap_abelhas);
            printf("{Capacidade de mel}: ");
            scanf("%d", &cap_mel);
            getchar();
            criar_colmeia(nome, cap_abelhas, cap_mel);
        } else if (escolha == 2) {
            char nome[100];
            printf("{Nome do apicultor}: ");
            fgets(nome, sizeof(nome), stdin);
            nome[strcspn(nome, "\n")] = 0;
            listar_colmeias(nome);
        } else if (escolha == 3) {
            int id_colmeia, quantidade;
            char rainha[10];
            bool rainha_presente;
            printf("{ID da colmeia}: ");
            scanf("%d", &id_colmeia);
            printf("{Quantidade de abelhas}: ");
            scanf("%d", &quantidade);
            getchar();
            printf("{Rainha presente? (true/false)}: ");
            fgets(rainha, sizeof(rainha), stdin);
            rainha[strcspn(rainha, "\n")] = 0;
            rainha_presente = (strcmp(rainha, "true") == 0);
            adicionar_abelha(id_colmeia, quantidade, rainha_presente);
        } else if (escolha == 4) {
            int id_colmeia;
            printf("{ID da colmeia}: ");
            scanf("%d", &id_colmeia);
            getchar();
            remover_colmeia(id_colmeia);
        } else if (escolha == 0) {
            printf("{Saindo...}\n");
            break;
        } else {
            printf("{Opção inválida!}\n");
        }
    }

    curl_global_cleanup();
    return 0;
}